File: services/intelligence/app/contracts/invariants.py
"""Contract invariants (docs/intelligence-engine/api-contract.md). Enforced on every result before it
leaves the engine and re-used by the tests on every fixture. Returns a list of violations (empty = ok)."""

from __future__ import annotations

from collections.abc import Iterable

from app.contracts.result import ResearchResult


def check_invariants(result: ResearchResult, fetched_urls: Iterable[str] | None = None) -> list[str]:
    problems: list[str] = []
    evidence = {e.id: e for e in result.evidence}
    if len(evidence) != len(result.evidence):
        problems.append("duplicate evidence ids")

    def refs(owner: str, ids: list[str]) -> None:
        for eid in ids:
            if eid not in evidence:
                problems.append(f"{owner}: evidence id {eid!r} does not resolve")

    for field in result.company.fields:
        refs(f"company.fields[{field.field}]", field.evidence_ids)
        if not field.evidence_ids:
            problems.append(f"company.fields[{field.field}]: field without evidence")
    for p in result.people:
        refs(f"people[{p.name}]", p.evidence_ids)
        if not p.evidence_ids:
            problems.append(f"people[{p.name}]: person without evidence")
    for s in result.signals:
        refs(f"signals[{s.id}]", s.evidence_ids)
        if not s.evidence_ids:
            problems.append(f"signals[{s.id}]: signal without evidence")
        if s.verified and not all(evidence[e].verification.verified for e in s.evidence_ids if e in evidence):
            problems.append(f"signals[{s.id}]: verified signal references unverified evidence")

    if fetched_urls is not None:
        fetched = set(fetched_urls)
        for e in result.evidence:
            if e.source_url not in fetched:
                problems.append(f"evidence[{e.id}]: source_url was not fetched by the engine")

    # Outreach narrative may only rest on verified evidence.
    refs("outreach", result.outreach.evidence_ids)
    for eid in result.outreach.evidence_ids:
        if eid in evidence and not evidence[eid].verification.verified:
            problems.append(f"outreach: references unverified evidence {eid!r}")

    # Scores may only rest on verified evidence.
    for item in result.icp.breakdown:
        refs(f"icp[{item.criterion}]", item.evidence_ids)
        for eid in item.evidence_ids:
            if eid in evidence and not evidence[eid].verification.verified and item.points > 0:
                problems.append(f"icp[{item.criterion}]: points awarded from unverified evidence {eid!r}")
    unverified_signal_ids = {s.id for s in result.signals if not s.verified}
    for it in result.intent.breakdown:
        if it.signal_id in unverified_signal_ids and it.points > 0:
            problems.append(f"intent: points awarded from unverified signal {it.signal_id!r}")
    for w in result.why_fit:
        refs(f"why_fit[{w.criterion}]", w.evidence_ids)
    return problems

File: lib/intelligence/invariants.ts
import type { EngineResearchResult } from "./schemas";

/**
 * Defensive re-check of the engine's contract invariants (mirror of services/intelligence/app/contracts/invariants.py).
 * The engine enforces them too; we never rely on that. A payload that violates any of them is QUARANTINED:
 * nothing from it is written. Returns the list of violations (empty = safe to persist).
 *
 * Rules: every evidence reference resolves; every claim-bearing item has evidence; a verified signal rests only on
 * verified evidence; the outreach narrative and every score point rest only on VERIFIED evidence/signals.
 */
export function checkInvariants(result: EngineResearchResult): string[] {
  const problems: string[] = [];
  const evidence = new Map(result.evidence.map((e) => [e.id, e]));
  if (evidence.size !== result.evidence.length) problems.push("duplicate evidence ids");

  const resolves = (owner: string, ids: string[]) => {
    for (const id of ids) if (!evidence.has(id)) problems.push(`${owner}: evidence id "${id}" does not resolve`);
  };

  for (const f of result.company.fields) {
    resolves(`company.fields[${f.field}]`, f.evidence_ids);
    if (f.evidence_ids.length === 0) problems.push(`company.fields[${f.field}]: field without evidence`);
    else if (!f.evidence_ids.every((id) => evidence.get(id)?.verification.verified)) problems.push(`company.fields[${f.field}]: rests on unverified evidence`);
  }
  for (const p of result.people) {
    resolves(`people[${p.name}]`, p.evidence_ids);
    if (p.evidence_ids.length === 0) problems.push(`people[${p.name}]: person without evidence`);
  }

  const signalIds = new Set<string>();
  for (const s of result.signals) {
    if (signalIds.has(s.id)) problems.push(`duplicate signal id ${s.id}`);
    signalIds.add(s.id);
    resolves(`signals[${s.id}]`, s.evidence_ids);
    if (s.evidence_ids.length === 0) problems.push(`signals[${s.id}]: signal without evidence`);
    if (s.verified && !s.evidence_ids.every((id) => evidence.get(id)?.verification.verified)) {
      problems.push(`signals[${s.id}]: verified signal references unverified evidence`);
    }
  }

  resolves("outreach", result.outreach.evidence_ids);
  for (const id of result.outreach.evidence_ids) {
    if (evidence.has(id) && !evidence.get(id)!.verification.verified) problems.push(`outreach: references unverified evidence "${id}"`);
  }

  for (const item of result.icp.breakdown) {
    resolves(`icp[${item.criterion}]`, item.evidence_ids);
    if (item.points > 0 && item.evidence_ids.some((id) => evidence.get(id) && !evidence.get(id)!.verification.verified)) {
      problems.push(`icp[${item.criterion}]: points awarded from unverified evidence`);
    }
  }
  const unverified = new Set(result.signals.filter((s) => !s.verified).map((s) => s.id));
  for (const i of result.intent.breakdown) {
    if (i.signal_id && unverified.has(i.signal_id) && i.points > 0) problems.push(`intent: points awarded from unverified signal "${i.signal_id}"`);
    if (i.signal_id && !signalIds.has(i.signal_id)) problems.push(`intent: signal "${i.signal_id}" does not exist`);
  }
  for (const w of result.why_fit) resolves(`why_fit[${w.criterion}]`, w.evidence_ids);
  return problems;
}
